### ISSUE TITLE
Sitemap generation is less stable than it used to be

### ISSUE THREAD

**Issue by stevejalim** [mozmeao/www-sitemap-generator#9]
For a long time, this tool was rock solid. Now, we find occasionally that the (locally run) bedrock app that is the core of the tool will return a 500 for a view, causing the entire run to fail.

Thinking about this, I'm wondering if it's because granian is running _too_ fast - something we've seen in production, too. 

Might be worth experimenting and setting `GRANIAN_WORKERS=1` and `GRANIAN_BLOCKING_THREADS=1` as env vars to slow things down (relatively, but still things will be fast enough)

**Comment by gi0baro** [comment]
@stevejalim just a note, if you want to limit the concurrency in Granian, it's definitely better to set `GRANIAN_BACKPRESSURE=N` where `N` is the maximum number of concurrent requests you want, rather than using blocking threads or workers to achieve that.

**Comment by stevejalim** [comment]
Thanks @gi0baro - good point!  We've used `GRANIAN_BACKPRESSURE` elsewhere and yep, that's a more natural way to throttle. We should try it in this situation.

--

Some tangential background: `www-sitemap-generator` uses the same `bedrock` Docker image that serves our main website. 

In this Issue, I defaulted to using blocking workers because of other experience we'd had with needing to throttle:  when the `bedrock` container is hooked up to postgres and we're not blocking at all, it was too easy to max out our available DB connections when the service went under load. 

Tuning `backpressure` to low enough numbers to reliably avoid saturating the DB resulted in lower RPS than blocking, based on some manual experiments. However, I kind of feel like maybe I am just Doing It Wrong and there's an angle I'm missing - e.g. is connection pooling the best way to tackle that, rather than throttling? 

This is kind of a tangent to the main issue, but I wanted to explain why it looks like I was suggesting putting training wheels on a superbike :) Also, I'd definitely welcome your take on this, if you have headspace. Shall I open an issue in the granian repo?

**Comment by gi0baro** [comment]
> In this Issue, I defaulted to using blocking workers because of other experience we'd had with needing to throttle: when the `bedrock` container is hooked up to postgres and we're not blocking at all, it was too easy to max out our available DB connections when the service went under load.
> 
> Tuning `backpressure` to low enough numbers to reliably avoid saturating the DB resulted in lower RPS than blocking, based on some manual experiments. However, I kind of feel like maybe I am just Doing It Wrong and there's an angle I'm missing - e.g. is connection pooling the best way to tackle that, rather than throttling?

Yeah , synchronous code and throttling are *quite a complicated thing*, as – when compared to async code – there's not a lot of ways for Granian to control what's going on under the hood.
Backpressure is the default way to limit concurrency, but that's handled on the TCP accept loop, so for instance if you have kept-alive connections, those wouldn't affect the counter; setting a limit to blocking threads is instead a hard limit: when the request needs to be forwarded to Python code it would just wait until a thread is free to accept that.
Connection pooling in synchronous code might be complicated as well, so I would say if you found that setting a hard limit to blocking threads works the best for you, it's absolutely fine.

> This is kind of a tangent to the main issue, but I wanted to explain why it looks like I was suggesting putting training wheels on a superbike :) Also, I'd definitely welcome your take on this, if you have headspace. Shall I open an issue in the granian repo?

Thank you for the additional context, I appreciate.
Issues/Discussions on Granian repo are always welcome, especially for having a broader context on real-world examples, so feel absolutely free to share your thoughts/experiments/feedbacks :)

**Comment by stevejalim** [comment]
Thanks @gi0baro - I appreciate your thoughtful reply.

Looking more into closed issues on the granian repo, I noticed that https://github.com/emmett-framework/granian/issues/324 sounds similar to what we originally encountered: Django opening far more DB connections that should logically be needed for low `backpressure` and `worker` settings. If/when we work out more about what's happening, I'll resume the conversation there and share what we discover.

**Comment by stevejalim** [comment]
Before the weekend I tweaked it to 
```
GRANIAN_WORKERS=4
GRANIAN_BLOCKING_THREADS=4
```
based on the idea that we were OOMing because we were trying to do too much at once (or, perhaps more likely, there's a memory leak in Bedrock that becomes v apparent under massive load. There's a separate conversation going on about profiing Bedrock for memory leaks)

We kept `GRANIAN_BACKPRESSURE=64` though I think that's possibly irrelevant with the blocking threads now anyway.

Generation has been stable all weekend, and actually pretty snappy.
